File: heaps/largestStoneWeight.py
# https://leetcode.com/problems/last-stone-weight/description/

# Approach: 
# 1. Sort the stones in ascending order.
# 2. Keep popping the last two elements from the list and find the difference.
# 3. If the difference is not zero, then add the difference to the list.
# 4. Sort the list again.
# 5. Repeat the process until the length of the list is 1.
# 6. Return the last element of the list if the list is not empty, else return 0.

def lastStoneWeight(stones):
    stones.sort()
    while len(stones) > 1:
        y = stones.pop()
        x = stones.pop()
        if x != y:
            stones.append(y - x)
            stones.sort()
    return stones[0] if stones else 0

File: heaps/largestStoneWeight.ts
// https://leetcode.com/problems/last-stone-weight/description/

// Approach:
// 1. Sort the stones in ascending order.
// 2. While there are more than 1 stones, do the following:
// 3. Get the last and second last stone.
// 4. Calculate the difference between the two.
// 5. Remove the last two stones.
// 6. If the difference is not 0, push the difference into the array.
// 7. Sort the array again.
// 8. Return the last stone if there is one, else return 0.

function lastStoneWeight(stones: number[]): number {
  if (stones.length === 1) return stones[0];
  if (!stones.length) return 0;

  stones.sort((a, b) => a - b);

  while (stones.length > 1) {
    const lastStone = stones[stones.length - 1];
    const secondLastStone = stones[stones.length - 2];
    const diff = lastStone - secondLastStone;

    stones.pop();
    stones.pop();

    if (diff) {
      stones.push(diff);
      stones.sort((a, b) => a - b);
    }
  }

  return stones.length ? stones[0] : 0;
}
